op-dripper/dripper/driver.go

```diff
@@ -0,0 +1,251 @@
+package dripper
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-dripper/bindings"
+	"github.com/ethereum-optimism/optimism/op-dripper/metrics"
+	"github.com/ethereum-optimism/optimism/op-service/txmgr"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type Client interface {
+	CodeAt(ctx context.Context, contract common.Address, blockNumber *big.Int) ([]byte, error)
+	CallContract(ctx context.Context, call ethereum.CallMsg, blockNumber *big.Int) ([]byte, error)
+}
+
+type DrippieContract interface {
+	GetDripCount(*bind.CallOpts) (*big.Int, error)
+	Created(*bind.CallOpts, *big.Int) (string, error)
+	Executable(*bind.CallOpts, string) (bool, error)
+}
+
+type DriverSetup struct {
+	Log    log.Logger
+	Metr   metrics.Metricer
+	Cfg    DripExecutorConfig
+	Txmgr  txmgr.TxManager
+	Client Client
+}
+
+type DripExecutor struct {
+	DriverSetup
+
+	wg   sync.WaitGroup
+	done chan struct{}
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	mutex   sync.Mutex
+	running bool
+
+	drippieContract DrippieContract
+	drippieABI      *abi.ABI
+}
+
+func NewDripExecutor(setup DriverSetup) (_ *DripExecutor, err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Ensure context does't leak.
+	defer func() {
+		if err != nil || recover() != nil {
+			cancel()
+		}
+	}()
+
+	if setup.Cfg.DrippieAddr == nil {
+		return nil, errors.New("drippie address is required")
+	}
+
+	return newDripExecutor(ctx, cancel, setup)
+}
+
+func newDripExecutor(ctx context.Context, cancel context.CancelFunc, setup DriverSetup) (*DripExecutor, error) {
+	drippieContract, err := bindings.NewDrippieCaller(*setup.Cfg.DrippieAddr, setup.Client)
+	if err != nil {
+		cancel()
+		return nil, fmt.Errorf("failed to create drippie at address %s: %w", setup.Cfg.DrippieAddr, err)
+	}
+
+	log.Info("connected to drippie", "address", setup.Cfg.DrippieAddr)
+	parsed, err := bindings.DrippieMetaData.GetAbi()
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	return &DripExecutor{
```

Comment:
I think you should set the context here so you don't have to do this `cancel()` and `defer func() { recover().. }` thing anymore.